examples/llm/inference/inference.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2022 MosaicML Examples authors
+# SPDX-License-Identifier: Apache-2.0
+import warnings
+from typing import List
+
+import torch
+from composer.core import get_precision_context
+from omegaconf import OmegaConf as om
+
+from examples.llm.src import COMPOSER_MODEL_REGISTRY, TOKENIZER_REGISTRY
+
+
+def sample_top_p(probs, p):
+    probs_sort, probs_idx = torch.sort(probs, dim=-1, descending=True)
+    probs_sum = torch.cumsum(probs_sort, dim=-1)
+    mask = probs_sum - probs_sort > p
+    probs_sort[mask] = 0.0
+    probs_sort.div_(probs_sort.sum(dim=-1, keepdim=True))
+    next_token = torch.multinomial(probs_sort, num_samples=1)
+    next_token = torch.gather(probs_idx, -1, next_token)
+    return next_token
+
+
+class MosaicGPTInference:
+
+    def __init__(self, cfg, model, tokenizer):
+        self.cfg = cfg
+        self.model = model
+        self.tokenizer = tokenizer
+
+    def generate(
+        self,
+        prompts: List[str],
+        max_gen_len: int,
+        temperature: float = 0.8,
+        top_p: float = 0.95,
+    ) -> List[str]:
+        bsz = len(prompts)
+
+        prompt_tokens = [self.tokenizer.encode(x) for x in prompts]
+
+        min_prompt_size = min([len(t) for t in prompt_tokens])
+        max_prompt_size = max([len(t) for t in prompt_tokens])
+
+        total_len = min(self.cfg.max_seq_len, max_gen_len + max_prompt_size)
+
+        tokens = torch.full((bsz, total_len), -100).cuda().long()
+        for k, t in enumerate(prompt_tokens):
+            tokens[k, :len(t)] = torch.tensor(t).long()
+        input_text_mask = tokens != -100
+        start_pos = min_prompt_size
+        for cur_pos in range(start_pos, total_len):
+            with torch.no_grad():
+                with get_precision_context(self.cfg.get('precision',
+                                                        'amp_bf16')):
+                    logits = self.model.forward(
+                        {'input_ids': tokens[:, :cur_pos]})
+            logits = logits[:, -1, :]
+            logits = logits.float()
+            if temperature > 0:
+                probs = torch.softmax(logits / temperature, dim=-1)
+                next_token = sample_top_p(probs, top_p)
+            else:
+                next_token = torch.argmax(logits, dim=-1)
+            next_token = next_token.reshape(-1)
+            # only replace token if prompt has already been generated
+            next_token = torch.where(input_text_mask[:, cur_pos],
+                                     tokens[:, cur_pos], next_token)
+            tokens[:, cur_pos] = next_token
```

Comment:
do we want to break out of the `for` loop if `next_token == self.tokenizer.eos_token_id` here?

Reply:
Oops, I was only thinking about batch size = 1 here

Reply:
and my other point, about deeper integration, will be resolved with Daniel's work. So I am resolving this